File: sims/src/cpu/strategies/override-cpu.ts
import { Hex } from '../hex';
import { SWITCH_MOVE_INDEX } from '../constants';
import {
  cpuActiveMonIndex,
  oppActiveMonIndex,
  monMaxHp,
  monCurrentHp,
  monCurrentStamina,
  monCurrentSpeed,
} from '../engine-view';
import { forkMeasureIncomingDamage } from '../heuristic-native';
import { BattleView } from '../battle-view';
import { BaseCpuStrategy, CpuMove, PlayerMove, StrategyState } from '../strategy';
import { HardCpu } from './hard-cpu';

/**
 * OVERRIDE CPU — a scripted-plan pilot for validating setup / resource / self-sacrifice lines the
 * heuristic pilots refuse to play ("unpiloted is not weak"). It wraps {@link HardCpu}: for the active
 * mon it consults a per-mon script (ordered rules with `when` / `once` / `maxUses` gates); the first
 * rule whose move is affordable this turn and whose gate passes is played, otherwise it delegates the
 * whole decision to hard. Reusable beyond the design pass — any "force this mon to play this line"
 * experiment registers a script here.
 *
 * Mons are keyed by BASE HP, which is unique per mon (see drool/mons.csv), because the battle view
 * exposes stats, not mon ids.
 *
 *   getCpuStrategy('override')   // then run via sim-tests/arena/*.ts --strategies override
 */

const CPU = 1n; // the CPU is always p1
const OPP = 0n;

/** Read-only position facts a `when` predicate can gate on. */
export interface OverrideCtx {
  /** CPU active mon current HP / max HP, in [0, 1]. */
  hpFrac: number;
  /** CPU active mon current stamina. */
  stamina: number;
  /** CPU active mon outspeeds the opponent's active mon. */
  outspeeds: boolean;
  /** The opponent's revealed move this turn would KO the CPU active mon if it stays in. */
  incomingLethal: boolean;
  /** The opponent's revealed move index (SWITCH_MOVE_INDEX / NO_OP when it isn't attacking). */
  oppMoveIndex: number;
  /** Turns elapsed this game. */
  turn: number;
  /** How many times THIS rule has already fired this game. */
  uses: number;
}

export interface OverrideRule {
  /** Move slot 0-3 to play. */
  move: number;
  /** Optional gate; the rule only fires when this returns true (or when omitted). */
  when?: (c: OverrideCtx) => boolean;
  /** Fire at most once per game. */
  once?: boolean;
  /** Fire at most N times per game. */
  maxUses?: number;
  /** Force a specific extraData target; defaults to the engine-picked valid target. */
  extraData?: number;
  /** Human label for logs. */
  label?: string;
}

/** An ordered list of rules; first affordable + passing rule wins, else fall through to hard. */
export type OverrideScript = OverrideRule[];

/**
 * Scripts keyed by base HP. Add a mon by dropping its base HP (from mons.csv) and an ordered rule
 * list. Start small — one signature line per mon is enough to test the hypothesis.
 */
export const OVERRIDE_SCRIPTS: Record<number, OverrideScript> = {
  // Ghouliath (baseHp 303): fire Eternal Grudge (slot 0) on the turn it would otherwise be KO'd,
  // while Rise From The Grave can still refund the self-KO. Once per game (the revive is one charge).
  303: [{ move: 0, when: (c) => c.incomingLethal, once: true, label: 'Eternal Grudge on lethal' }],

  // Aurox (baseHp 400): the tank line — Iron Wall (slot 2) on a fresh, stamina-flush entry, then
  // Bull Rush (slot 3) as the default so Up Only ramps behind the regen.
  400: [
    { move: 2, when: (c) => c.hpFrac > 0.9 && c.stamina >= 3, maxUses: 1, label: 'Iron Wall on entry' },
    { move: 3, label: 'Bull Rush' },
  ],

  // Inutia (baseHp 351): the weave-and-pass line — Initialize (slot 1) on a fresh, safe entry, then
  // Hit and Dip (slot 3) to pivot and pass the boost. NOTE: the pivot target is engine-picked, not
  // frailest-partner-aware, so this tests "does weaving beat mono-attacking," not a hand-picked pass.
  351: [
    { move: 1, when: (c) => c.hpFrac > 0.75, once: true, label: 'Initialize on entry' },
    { move: 3, once: true, label: 'Hit and Dip pivot (pass the boost)' },
  ],
};

export class OverrideCpu extends BaseCpuStrategy {
  readonly name = 'override';
  private readonly fallback = new HardCpu();

  createState(): StrategyState {
    return { fallback: this.fallback.createState(), turn: 0, uses: {} };
  }

  decide(view: BattleView, playerMove: PlayerMove, rng: () => number, state: StrategyState): CpuMove {
    const e = view.engine;
    const bk = view.bk as Hex;
    state.turn = (state.turn as number) + 1;

    const activeIdx = cpuActiveMonIndex(e, bk);
    const script = OVERRIDE_SCRIPTS[monMaxHp(e, bk, CPU, activeIdx)];

    if (script) {
      const { moves } = this.validMoves(e, bk, rng); // affordable slots, each with a valid target
      const uses = state.uses as Record<string, number>;
      let ctx: OverrideCtx | undefined;

      for (let i = 0; i < script.length; i++) {
        const rule = script[i];
        const key = `${activeIdx}:${i}`;
        const fired = uses[key] ?? 0;
        if (rule.once && fired >= 1) continue;
        if (rule.maxUses !== undefined && fired >= rule.maxUses) continue;

        const affordable = moves.find((m) => m.moveIndex === rule.move);
        if (!affordable) continue; // not castable this turn (stamina / forced-switch / etc.)

        if (rule.when) {
          if (!ctx) ctx = buildCtx(e, bk, activeIdx, playerMove, state.turn as number);
          if (!rule.when({ ...ctx, uses: fired })) continue;
        }

        uses[key] = fired + 1;
        return { moveIndex: affordable.moveIndex, extraData: rule.extraData ?? affordable.extraData };
      }
    }

    return this.fallback.decide(view, playerMove, rng, state.fallback as StrategyState);
  }
}

function buildCtx(e: any, bk: Hex, activeIdx: number, playerMove: PlayerMove, turn: number): OverrideCtx {
  const curHp = monCurrentHp(e, bk, CPU, activeIdx);
  const maxHp = monMaxHp(e, bk, CPU, activeIdx);
  const oppIdx = oppActiveMonIndex(e, bk);

  // Incoming lethal: only meaningful when the opponent actually attacks this turn.
  let incomingLethal = false;
  if (playerMove.moveIndex < SWITCH_MOVE_INDEX) {
    try {
      incomingLethal =
        forkMeasureIncomingDamage(e, bk, playerMove.moveIndex, playerMove.extraData, null, 0n) >= curHp;
    } catch {
      incomingLethal = false; // a fork that fails on an odd state just reads as non-lethal
    }
  }

  return {
    hpFrac: maxHp > 0 ? curHp / maxHp : 0,
    stamina: monCurrentStamina(e, bk, CPU, activeIdx),
    outspeeds: monCurrentSpeed(e, bk, CPU, activeIdx) > monCurrentSpeed(e, bk, OPP, oppIdx),
    incomingLethal,
    oppMoveIndex: playerMove.moveIndex,
    turn,
    uses: 0,
  };
}

File: transpiler/strategies-rs/src/override_cpu.rs
//! OVERRIDE CPU — port of `sims/src/cpu/strategies/override-cpu.ts`: a
//! scripted-plan pilot wrapping the hard CPU. For the active mon it
//! consults a per-mon script (ordered rules with `when` / max-uses
//! gates); the first rule whose move is affordable this turn and whose
//! gate passes is played, otherwise the whole decision delegates to hard.
//!
//! Mons are keyed by BASE HP (unique per mon, see drool/mons.csv).
//!
//! Parity notes: on a script-mon turn the candidate enumeration runs
//! HERE and — when no rule fires — AGAIN inside the hard fallback,
//! consuming the rng stream twice, exactly like the TS wrapper. The
//! `incomingLethal` fact forks with salt 0 and reads false if the fork
//! panics (the TS try/catch).

use std::collections::HashMap;
use std::panic::{catch_unwind, AssertUnwindSafe};

use crate::hard::{self, HardState};
use crate::jsrng::JsRng;
use crate::native::{fork_measure_incoming_damage, ForkCache};
use crate::sim::Sim;
use crate::view::{
    calculate_valid_moves, mon_current_hp, mon_current_speed, mon_current_stamina, mon_max_hp,
    BattleView, Mv, Seat, SWITCH_MOVE_INDEX, VCPU, VOPP,
};

/// Read-only position facts a `when` predicate can gate on.
#[derive(Clone, Copy)]
pub struct OverrideCtx {
    /// CPU active mon current HP / max HP, in [0, 1].
    pub hp_frac: f64,
    /// CPU active mon current stamina.
    pub stamina: i64,
    /// CPU active mon outspeeds the opponent's active mon.
    pub outspeeds: bool,
    /// The opponent's revealed move this turn would KO the CPU active mon.
    pub incoming_lethal: bool,
    /// The opponent's revealed move index (SWITCH/NO_OP when not attacking).
    pub opp_move_index: u8,
    /// Turns elapsed this game.
    pub turn: u32,
    /// How many times THIS rule has already fired this game.
    pub uses: u32,
}

pub struct OverrideRule {
    /// Move slot 0-3 to play.
    pub mv: u8,
    /// Optional gate; the rule only fires when this returns true.
    pub when: Option<fn(&OverrideCtx) -> bool>,
    /// Fire at most N times per game (TS `once` == Some(1)).
    pub max_uses: Option<u32>,
    /// Force a specific extraData target; None = the engine-picked target.
    pub extra_data: Option<u16>,
    /// Human label for logs / doc parity with the TS scripts.
    pub label: &'static str,
}

const fn rule(mv: u8, when: Option<fn(&OverrideCtx) -> bool>, max_uses: Option<u32>, label: &'static str) -> OverrideRule {
    OverrideRule { mv, when, max_uses, extra_data: None, label }
}

fn aurox_iron_wall(c: &OverrideCtx) -> bool {
    c.hp_frac > 0.9 && c.stamina >= 3
}
fn under_half_hp(c: &OverrideCtx) -> bool {
    c.hp_frac < 0.5
}
fn stamina_ge_3(c: &OverrideCtx) -> bool {
    c.stamina >= 3
}
fn stamina_lt_3(c: &OverrideCtx) -> bool {
    c.stamina < 3
}

/// Scripts keyed by base HP — kept rule-for-rule identical to
/// `OVERRIDE_SCRIPTS` in the TS registry (same ordering, gates, labels).
fn script_for(base_hp: i64) -> Option<&'static [OverrideRule]> {
    // Aurox: the tank line — Iron Wall on a fresh, stamina-flush entry,
    // then Bull Rush as the default so Up Only ramps behind the regen.
    static AUROX: &[OverrideRule] = &[
        rule(2, Some(aurox_iron_wall), Some(1), "Iron Wall on entry"),
        rule(3, None, None, "Bull Rush"),
    ];
    // Iblivion: the turn-1 Loop line — Loop once on entry, Brightback
    // sustain under half HP, Unbounded Strike as the default.
    static IBLIVION: &[OverrideRule] = &[
        rule(1, None, Some(1), "Loop on entry (+15%)"),
        rule(2, Some(under_half_hp), None, "Brightback sustain"),
        rule(0, None, None, "Unbounded Strike"),
    ];
    // Volthare: prefer Mega Star Blast whenever affordable.
    static VOLTHARE: &[OverrideRule] =
        &[rule(2, Some(stamina_ge_3), None, "Mega Star Blast preference")];
    // Embursa: arm Q5 once on the first action, then fall through to hard.
    static EMBURSA: &[OverrideRule] = &[rule(3, None, Some(1), "Arm Q5")];
    // Pengym: the committed Frostbite combo — tag, Deep Freeze while
    // affordable, re-tag while saving stamina.
    static PENGYM: &[OverrideRule] = &[
        rule(0, None, Some(1), "Chill Out tag"),
        rule(2, Some(stamina_ge_3), Some(2), "Deep Freeze"),
        rule(0, Some(stamina_lt_3), Some(2), "Chill Out while saving"),
    ];
    // Inutia: arm Chain Expansion once, hard for everything else.
    static INUTIA: &[OverrideRule] = &[rule(0, None, Some(1), "Arm Chain Expansion")];

    match base_hp {
        400 => Some(AUROX),
        277 => Some(IBLIVION),
        310 => Some(VOLTHARE),
        420 => Some(EMBURSA),
        371 => Some(PENGYM),
        351 => Some(INUTIA),
        _ => None,
    }
}

#[derive(Default)]
pub struct OverrideState {
    pub turn: u32,
    /// Fire counts keyed by (active mon slot, rule index) — per game.
    pub uses: HashMap<(usize, usize), u32>,
    pub fallback: HardState,
}

fn build_ctx(
    sim: &mut Sim,
    seat: Seat,
    view: &BattleView,
    active_idx: usize,
    pm: Mv,
    turn: u32,
) -> OverrideCtx {
    let bk = view.bk;
    let cur_hp = mon_current_hp(sim, seat, bk, VCPU, active_idx);
    let max_hp = mon_max_hp(sim, seat, bk, VCPU, active_idx);
    let opp_idx = view.opp_active;

    // Incoming lethal: only meaningful when the opponent actually attacks;
    // a fork that fails on an odd state just reads as non-lethal. (A panic
    // mid-fork can strand that fork's cloned maps until the per-game world
    // drops — bounded, and the TS reference leaks the same way.)
    let mut incoming_lethal = false;
    if pm.move_index < SWITCH_MOVE_INDEX {
        let mut fc = ForkCache::new();
        incoming_lethal = catch_unwind(AssertUnwindSafe(|| {
            fork_measure_incoming_damage(sim, seat, &mut fc, pm.move_index, pm.extra_data, None, 0) >= cur_hp
        }))
        .unwrap_or(false);
        fc.dispose_all(sim);
    }

    OverrideCtx {
        hp_frac: if max_hp > 0 { cur_hp as f64 / max_hp as f64 } else { 0.0 },
        stamina: mon_current_stamina(sim, seat, bk, VCPU, active_idx),
        outspeeds: mon_current_speed(sim, seat, bk, VCPU, active_idx)
            > mon_current_speed(sim, seat, bk, VOPP, opp_idx),
        incoming_lethal,
        opp_move_index: pm.move_index,
        turn,
        uses: 0,
    }
}

pub fn decide(
    sim: &mut Sim,
    seat: Seat,
    view: &BattleView,
    pm: Mv,
    rng: &mut JsRng,
    st: &mut OverrideState,
) -> Mv {
    let bk = view.bk;
    st.turn += 1;

    let active_idx = view.cpu_active;
    if let Some(script) = script_for(mon_max_hp(sim, seat, bk, VCPU, active_idx)) {
        // Affordable slots, each with a valid target (rng draws here AND
        // again in the fallback — the TS double-enumeration).
        let valid = calculate_valid_moves(sim, seat, bk, rng);
        let mut ctx: Option<OverrideCtx> = None;

        for (i, r) in script.iter().enumerate() {
            let key = (active_idx, i);
            let fired = *st.uses.get(&key).unwrap_or(&0);
            if let Some(max) = r.max_uses {
                if fired >= max {
                    continue;
                }
            }
            let Some(affordable) = valid.moves.iter().find(|m| m.move_index == r.mv) else {
                continue; // not castable this turn (stamina / forced-switch / etc.)
            };
            if let Some(pred) = r.when {
                if ctx.is_none() {
                    ctx = Some(build_ctx(sim, seat, view, active_idx, pm, st.turn));
                }
                let mut c = ctx.unwrap();
                c.uses = fired;
                if !pred(&c) {
                    continue;
                }
            }
            st.uses.insert(key, fired + 1);
            return Mv {
                move_index: affordable.move_index,
                extra_data: r.extra_data.unwrap_or(affordable.extra_data),
            };
        }
    }

    hard::decide(sim, seat, view, pm, rng, &mut st.fallback)
}
